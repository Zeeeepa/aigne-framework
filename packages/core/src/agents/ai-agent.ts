import { type ZodObject, type ZodType, z } from "zod";
import { DefaultMemory, type DefaultMemoryOptions } from "../memory/default-memory/index.js";
import { MemoryAgent } from "../memory/memory.js";
import { MESSAGE_KEY, PromptBuilder } from "../prompt/prompt-builder.js";
import { AgentMessageTemplate, ToolMessageTemplate } from "../prompt/template.js";
import { checkArguments, isEmpty } from "../utils/type-utils.js";
import {
  Agent,
  type AgentInvokeOptions,
  type AgentOptions,
  type AgentProcessAsyncGenerator,
  type AgentResponseStream,
  type Message,
  agentOptionsSchema,
} from "./agent.js";
import {
  ChatModel,
  type ChatModelInput,
  type ChatModelInputMessage,
  type ChatModelOutput,
  type ChatModelOutputToolCall,
} from "./chat-model.js";
import type { GuideRailAgentOutput } from "./guide-rail-agent.js";
import { isTransferAgentOutput } from "./types.js";

/**
 * Configuration options for an AI Agent
 *
 * These options extend the base agent options with AI-specific parameters
 * like model configuration, prompt instructions, and tool choice.
 *
 * @template I The input message type the agent accepts
 * @template O The output message type the agent returns
 */
export interface AIAgentOptions<I extends Message = Message, O extends Message = Message>
  extends Omit<AgentOptions<I, O>, "memory"> {
  /**
   * The language model to use for this agent
   *
   * If not provided, the agent will use the model from the context
   */
  model?: ChatModel;

  /**
   * Instructions to guide the AI model's behavior
   *
   * Can be a simple string or a full PromptBuilder instance for
   * more complex prompt templates
   */
  instructions?: string | PromptBuilder;

  /**
   * Custom key to use for text output in the response
   *
   * Defaults to $message if not specified
   */
  outputKey?: string;

  /**
   * Controls how the agent uses tools during execution
   *
   * @default AIAgentToolChoice.auto
   */
  toolChoice?: AIAgentToolChoice | Agent;

  /**
   * Whether to catch errors from tool execution and continue processing.
   * If set to false, the agent will throw an error if a tool fails.
   *
   * @default true
   */
  catchToolsError?: boolean;

  /**
   * Whether to include memory agents as tools for the AI model
   *
   * When set to true, memory agents will be made available as tools
   * that the model can call directly to retrieve or store information.
   * This enables the agent to explicitly interact with its memories.
   *
   * @default false
   */
  memoryAgentsAsTools?: boolean;

  /**
   * Custom prompt template for formatting memory content
   *
   * Allows customization of how memories are presented to the AI model.
   * If not provided, the default template from MEMORY_MESSAGE_TEMPLATE will be used.
   *
   * The template receives a {{memories}} variable containing serialized memory content.
   */
  memoryPromptTemplate?: string;

  memory?: AgentOptions<I, O>["memory"] | DefaultMemoryOptions | true;
}

/**
 * Tool choice options for AI agents
 *
 * Controls how the agent decides to use tools during execution
 */
export enum AIAgentToolChoice {
  /**
   * Let the model decide when to use tools
   */
  auto = "auto",

  /**
   * Disable tool usage
   */
  none = "none",

  /**
   * Force tool usage
   */
  required = "required",

  /**
   * Choose exactly one tool and route directly to it
   */
  router = "router",
}

/**
 * Zod schema for validating AIAgentToolChoice values
 *
 * Used to ensure that toolChoice receives valid values
 *
 * @hidden
 */
export const aiAgentToolChoiceSchema = z.union(
  [z.nativeEnum(AIAgentToolChoice), z.instanceof(Agent)],
  {
    message: `aiAgentToolChoice must be ${Object.values(AIAgentToolChoice).join(", ")}, or an Agent`,
  },
);

/**
 * Zod schema for validating AIAgentOptions
 *
 * Extends the base agent options schema with AI-specific parameters
 *
 * @hidden
 */
export const aiAgentOptionsSchema: ZodObject<{
  [key in keyof AIAgentOptions]: ZodType<AIAgentOptions[key]>;
}> = agentOptionsSchema.extend({
  model: z.instanceof(ChatModel).optional(),
  instructions: z.union([z.string(), z.instanceof(PromptBuilder)]).optional(),
  outputKey: z.string().optional(),
  toolChoice: aiAgentToolChoiceSchema.optional(),
  memoryAgentsAsTools: z.boolean().optional(),
  memoryPromptTemplate: z.string().optional(),
});

/**
 * AI-powered agent that leverages language models
 *
 * AIAgent connects to language models to process inputs and generate responses,
 * with support for streaming, function calling, and tool usage.
 *
 * Key features:
 * - Connect to any language model
 * - Use customizable instructions and prompts
 * - Execute tools/function calls
 * - Support streaming responses
 * - Router mode for specialized agents
 *
 * @template I The input message type the agent accepts
 * @template O The output message type the agent returns
 *
 * @example
 * Basic AIAgent creation:
 * {@includeCode ../../test/agents/ai-agent.test.ts#example-ai-agent-basic}
 */
export class AIAgent<I extends Message = Message, O extends Message = Message> extends Agent<I, O> {
  /**
   * Create an AIAgent with the specified options
   *
   * Factory method that provides a convenient way to create new AI agents
   *
   * @param options Configuration options for the AI agent
   * @returns A new AIAgent instance
   *
   * @example
   * AI agent with custom instructions:
   * {@includeCode ../../test/agents/ai-agent.test.ts#example-ai-agent-instructions}
   */
  static from<I extends Message, O extends Message>(options: AIAgentOptions<I, O>): AIAgent<I, O> {
    return new AIAgent(options);
  }

  /**
   * Create an AIAgent instance
   *
   * @param options Configuration options for the AI agent
   */
  constructor(options: AIAgentOptions<I, O>) {
    super({
      ...options,
      memory: !options.memory
        ? undefined
        : Array.isArray(options.memory) || options.memory instanceof MemoryAgent
          ? options.memory
          : new DefaultMemory(options.memory === true ? {} : options.memory),
    });
    checkArguments("AIAgent", aiAgentOptionsSchema, options);

    this.model = options.model;
    this.instructions =
      typeof options.instructions === "string"
        ? PromptBuilder.from(options.instructions)
        : (options.instructions ?? new PromptBuilder());
    this.outputKey = options.outputKey;
    this.toolChoice = options.toolChoice;
    this.memoryAgentsAsTools = options.memoryAgentsAsTools;
    this.memoryPromptTemplate = options.memoryPromptTemplate;

    if (typeof options.catchToolsError === "boolean")
      this.catchToolsError = options.catchToolsError;
  }

  /**
   * The language model used by this agent
   *
   * If not set on the agent, the model from the context will be used
   */
  model?: ChatModel;

  /**
   * Instructions for the language model
   *
   * Contains system messages, user templates, and other prompt elements
   * that guide the model's behavior
   *
   * @example
   * Custom prompt builder:
   * {@includeCode ../../test/agents/ai-agent.test.ts#example-ai-agent-prompt-builder}
   */
  instructions: PromptBuilder;

  /**
   * Custom key to use for text output in the response
   *
   * @example
   * Setting a custom output key:
   * {@includeCode ../../test/agents/ai-agent.test.ts#example-ai-agent-custom-output-key}
   */
  outputKey?: string;

  /**
   * Controls how the agent uses tools during execution
   *
   * @example
   * Automatic tool choice:
   * {@includeCode ../../test/agents/ai-agent.test.ts#example-ai-agent-tool-choice-auto}
   *
   * @example
   * Router tool choice:
   * {@includeCode ../../test/agents/ai-agent.test.ts#example-ai-agent-router}
   */
  toolChoice?: AIAgentToolChoice | Agent;

  /**
   * Whether to include memory agents as tools for the AI model
   *
   * When set to true, memory agents will be made available as tools
   * that the model can call directly to retrieve or store information.
   * This enables the agent to explicitly interact with its memories.
   */
  memoryAgentsAsTools?: boolean;

  /**
   * Custom prompt template for formatting memory content
   *
   * Allows customization of how memories are presented to the AI model.
   * If not provided, the default template from MEMORY_MESSAGE_TEMPLATE will be used.
   *
   * The template receives a {{memories}} variable containing serialized memory content.
   */
  memoryPromptTemplate?: string;

  /**
   * Whether to catch error from tool execution and continue processing.
   * If set to false, the agent will throw an error if a tool fails
   *
   * @default true
   */
  catchToolsError = true;

  /**
   * Process an input message and generate a response
   *
   * @protected
   */
  async *process(input: I, options: AgentInvokeOptions): AgentProcessAsyncGenerator<O> {
    const model = this.model ?? options.context.model;
    if (!model) throw new Error("model is required to run AIAgent");

    const { toolAgents, ...modelInput } = await this.instructions.build({
      agent: this,
      input,
      model,
      context: options.context,
    });

    const toolsMap = new Map<string, Agent>(toolAgents?.map((i) => [i.name, i]));

    if (this.toolChoice === "router") {
      yield* this._processRouter(input, model, modelInput, options, toolsMap);
      return;
    }

    const toolCallMessages: ChatModelInputMessage[] = [];
    const outputKey = this.outputKey || MESSAGE_KEY;

    for (;;) {
      const modelOutput: ChatModelOutput = {};

      const stream = await options.context.invoke(
        model,
        { ...modelInput, messages: modelInput.messages.concat(toolCallMessages) },
        { streaming: true },
      );

      for await (const value of stream) {
        if (value.delta.text?.text) {
          yield { delta: { text: { [outputKey]: value.delta.text.text } } };
        }

        if (value.delta.json) {
          Object.assign(modelOutput, value.delta.json);
        }
      }

      const { toolCalls, json, text } = modelOutput;

      if (toolCalls?.length) {
        const executedToolCalls: {
          call: ChatModelOutputToolCall;
          output: Message;
        }[] = [];

        // Execute tools
        for (const call of toolCalls) {
          const tool = toolsMap.get(call.function.name);
          if (!tool) throw new Error(`Tool not found: ${call.function.name}`);

          // NOTE: should pass both arguments (model generated) and input (user provided) to the tool
          const output = await this.invokeSkill(
            tool,
            { ...input, ...call.function.arguments },
            options,
          ).catch((error) => {
            if (!this.catchToolsError) {
              return Promise.reject(error);
            }

            return {
              isError: true,
              error: {
                message: error.message,
              },
            };
          });

          // NOTE: Return transfer output immediately
          if (isTransferAgentOutput(output)) {
            return output;
          }

          executedToolCalls.push({ call, output });
        }

        // Continue LLM function calling loop if any tools were executed
        if (executedToolCalls.length) {
          toolCallMessages.push(
            AgentMessageTemplate.from(
              undefined,
              executedToolCalls.map(({ call }) => call),
            ).format(),
            ...executedToolCalls.map(({ call, output }) =>
              ToolMessageTemplate.from(output, call.id).format(),
            ),
          );

          continue;
        }
      }

      const result = {} as O;

      if (json) {
        Object.assign(result, json);
      } else if (text) {
        Object.assign(result, { [outputKey]: text });
      }

      if (!isEmpty(result)) {
        yield { delta: { json: result } };
      }
      return;
    }
  }

  protected override async onGuideRailError(
    error: GuideRailAgentOutput,
  ): Promise<O | GuideRailAgentOutput> {
    const outputKey = this.outputKey || MESSAGE_KEY;
    return {
      [outputKey]: error.reason,
    };
  }

  /**
   * Process router mode requests
   *
   * In router mode, the agent sends a single request to the model to determine
   * which tool to use, then routes the request directly to that tool
   *
   * @protected
   */
  async *_processRouter(
    input: I,
    model: ChatModel,
    modelInput: ChatModelInput,
    options: AgentInvokeOptions,
    toolsMap: Map<string, Agent>,
  ): AgentProcessAsyncGenerator<O> {
    const {
      toolCalls: [call] = [],
    } = await options.context.invoke(model, modelInput);

    if (!call) {
      throw new Error("Router toolChoice requires exactly one tool to be executed");
    }

    const tool = toolsMap.get(call.function.name);
    if (!tool) throw new Error(`Tool not found: ${call.function.name}`);

    const stream = await options.context.invoke(
      tool,
      { ...call.function.arguments, ...input },
      { streaming: true, sourceAgent: this },
    );

    yield* stream as AgentResponseStream<O>;
  }
}
