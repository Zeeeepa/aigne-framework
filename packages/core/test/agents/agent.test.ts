import { expect, spyOn, test } from "bun:test";
import assert from "node:assert";
import { inspect } from "node:util";
import {
  Agent,
  type AgentInput,
  type AgentInvokeOptions,
  type AgentOutput,
  type AgentProcessAsyncGenerator,
  type AgentResponseChunk,
  type AgentResponseStream,
  AIAgent,
  AIAgentToolChoice,
  AIGNE,
  FunctionAgent,
  isAgentResponseDelta,
  type Message,
  textDelta,
} from "@aigne/core";
import { guideRailAgentOptions } from "@aigne/core/agents/guide-rail-agent";
import {
  readableStreamToArray,
  stringToAgentResponseStream,
} from "@aigne/core/utils/stream-utils.js";
import { z } from "zod";
import { OpenAIChatModel } from "../_mocks/mock-models.js";
import { expectType } from "../_utils/expect.js";
import { createToolCallResponse } from "../_utils/openai-like-utils.js";

test("Custom agent", async () => {
  // #region example-custom-agent

  class MyAgent extends Agent {
    process(input: Message): Message {
      console.log(input);

      return {
        text: "Hello, How can I assist you today?",
      };
    }
  }

  const agent = new MyAgent();

  const result = await agent.invoke({ message: "hello" });

  console.log(result); // { text: "Hello, How can I assist you today?" }

  expect(result).toEqual({ text: "Hello, How can I assist you today?" });

  // #endregion example-custom-agent
});

test("Custom agent returning a direct object", async () => {
  // #region example-process-direct-response

  class DirectResponseAgent extends Agent {
    process(input: Message): Message {
      // Process input and return a direct object response
      return {
        text: `Hello, I received your message: ${JSON.stringify(input)}`,
        confidence: 0.95,
        timestamp: new Date().toISOString(),
      };
    }
  }

  const agent = new DirectResponseAgent();

  const result = await agent.invoke({ message: "Hello" });

  console.log(result); // { text: "Hello, I received your message: { message: 'Hello' }", confidence: 0.95, timestamp: "2023-10-01T12:00:00Z" }

  expect(result).toEqual({
    text: expect.stringContaining("Hello, I received your message:"),
    confidence: 0.95,
    timestamp: expect.any(String),
  });

  // #endregion example-process-direct-response
});

test("Agent returning a ReadableStream", async () => {
  // #region example-process-streaming-response

  class StreamResponseAgent extends Agent {
    process(_input: Message): AgentResponseStream<Message> {
      // Return a ReadableStream as a streaming response
      return new ReadableStream({
        start(controller) {
          controller.enqueue(textDelta({ text: "Hello" }));
          controller.enqueue(textDelta({ text: ", " }));
          controller.enqueue(textDelta({ text: "This" }));
          controller.enqueue(textDelta({ text: " is" }));
          controller.enqueue(textDelta({ text: "..." }));
          controller.close();
        },
      });
    }
  }

  const agent = new StreamResponseAgent();
  const stream = await agent.invoke({ message: "Hello" }, { streaming: true });

  let fullText = "";
  for await (const chunk of stream) {
    if (isAgentResponseDelta(chunk)) {
      const text = chunk.delta.text?.text;
      if (text) fullText += text;
    }
  }

  console.log(fullText); // Output: "Hello, This is..."

  expect(fullText).toBe("Hello, This is...");
  // #endregion example-process-streaming-response
});

test("Agent using AsyncGenerator", async () => {
  // #region example-process-async-generator

  class AsyncGeneratorAgent extends Agent {
    async *process(
      _input: Message,
      _options: AgentInvokeOptions,
    ): AgentProcessAsyncGenerator<Message> {
      // Use async generator to produce streaming results
      yield textDelta({ message: "This" });
      yield textDelta({ message: "," });
      yield textDelta({ message: " " });
      yield textDelta({ message: "This" });
      yield textDelta({ message: " " });
      yield textDelta({ message: "is" });
      yield textDelta({ message: "..." });

      // Optional return a JSON object at the end
      return { time: new Date().toISOString() };
    }
  }

  const agent = new AsyncGeneratorAgent();
  const stream = await agent.invoke({ message: "Hello" }, { streaming: true });

  const message: string[] = [];
  let json: Message | undefined;

  for await (const chunk of stream) {
    if (isAgentResponseDelta(chunk)) {
      const text = chunk.delta.text?.message;
      if (text) message.push(text);
      if (chunk.delta.json) json = chunk.delta.json;
    }
  }

  console.log(message); // Output: ["This", ",", " ", "This", " ", "is", "..."]
  console.log(json); // Output: { time: "2023-10-01T12:00:00Z" }

  expect(message).toEqual(["This", ",", " ", "This", " ", "is", "..."]);
  expect(json).toEqual({ time: expect.any(String) });

  // #endregion example-process-async-generator
});

test("Agent returning another agent (transfer agent)", async () => {
  // #region example-process-transfer

  class SpecialistAgent extends Agent {
    process(_input: Message): Message {
      return {
        response: "This is a specialist response",
        expertise: "technical",
      };
    }
  }

  class MainAgent extends Agent {
    process(_input: Message): Agent {
      // Create a specialized agent for handling technical issues
      return new SpecialistAgent();
    }
  }

  const aigne = new AIGNE({});
  const mainAgent = new MainAgent();

  const result = await aigne.invoke(mainAgent, { message: "technical question" });
  console.log(result); // { response: "This is a specialist response", expertise: "technical" }

  expect(result).toEqual({ response: "This is a specialist response", expertise: "technical" });

  // #endregion example-process-transfer
});

test("Agent.invoke with regular response", async () => {
  // #region example-invoke

  // Create a chat model
  const model = new OpenAIChatModel();

  spyOn(model, "process").mockReturnValueOnce(
    Promise.resolve(stringToAgentResponseStream("Hello, How can I assist you today?")),
  );

  // AIGNE: Main execution engine of AIGNE Framework.
  const aigne = new AIGNE({
    model,
  });

  // Create an Agent instance
  const agent = AIAgent.from({
    name: "chat",
    description: "A chat agent",
    inputKey: "message",
  });

  // Invoke the agent
  const result = await aigne.invoke(agent, { message: "hello" });

  console.log(result); // Output: { message: "Hello, How can I assist you today?" }

  expect(result).toEqual({ message: "Hello, How can I assist you today?" });

  // #endregion example-invoke
});

test("Agent.invoke with streaming response", async () => {
  // #region example-invoke-streaming

  // Create a chat model
  const model = new OpenAIChatModel();

  spyOn(model, "process").mockReturnValueOnce(
    Promise.resolve(stringToAgentResponseStream("Hello, How can I assist you today?")),
  );

  // AIGNE: Main execution engine of AIGNE Framework.
  const aigne = new AIGNE({
    model,
  });

  // Create an Agent instance
  const agent = AIAgent.from({
    name: "chat",
    description: "A chat agent",
    inputKey: "message",
  });

  // Invoke the agent with streaming enabled
  const stream = await aigne.invoke(agent, { message: "hello" }, { streaming: true });

  const chunks: string[] = [];

  // Read the stream using an async iterator
  for await (const chunk of stream) {
    if (isAgentResponseDelta(chunk)) {
      const text = chunk.delta.text?.message;
      if (text) {
        chunks.push(text);
      }
    }
  }

  console.log(chunks); // Output: ["Hello", ",", " ", "How", " ", "can", " ", "I", " ", "assist", " ", "you", " ", "today", "?"]

  expect(chunks).toMatchSnapshot();

  // #endregion example-invoke-streaming
});

test("FunctionAgent.from a function", async () => {
  // #region example-function-agent

  const agent = FunctionAgent.from(({ name }: { name: string }) => {
    return {
      greeting: `Hello, ${name}!`,
    };
  });

  const result = await agent.invoke({ name: "Alice" });

  console.log(result); // Output: { greeting: "Hello, Alice!" }

  expect(result).toEqual({ greeting: "Hello, Alice!" });

  // #endregion example-function-agent
});

test("FunctionAgent.from a function without options", async () => {
  // #region example-function-agent-from-function

  const agent = FunctionAgent.from(({ a, b }: { a: number; b: number }) => {
    return { sum: a + b };
  });

  const result = await agent.invoke({ a: 5, b: 10 });

  console.log(result); // Output: { sum: 15 }

  expect(result).toEqual({ sum: 15 });

  // #endregion example-function-agent-from-function
});

test("FunctionAgent.from a function return stream", async () => {
  // #region example-function-agent-stream

  const agent = FunctionAgent.from(({ name }: { name: string }) => {
    return new ReadableStream<AgentResponseChunk<{ text: string }>>({
      start(controller) {
        controller.enqueue(textDelta({ text: "Hello" }));
        controller.enqueue(textDelta({ text: ", " }));
        controller.enqueue(textDelta({ text: name }));
        controller.enqueue(textDelta({ text: "!" }));
        controller.close();
      },
    });
  });

  const result = await agent.invoke({ name: "Alice" });

  console.log(result); // Output: { text: "Hello, Alice!" }

  expect(result).toEqual({ text: "Hello, Alice!" });

  // #endregion example-function-agent-stream
});

test("FunctionAgent.from a function return async iterator", async () => {
  // #region example-function-agent-async-generator

  const agent = FunctionAgent.from(async function* ({ name }: { name: string }) {
    yield textDelta({ text: "Hello" });
    yield textDelta({ text: ", " });
    yield textDelta({ text: name });
    yield textDelta({ text: "!" });
  });

  const result = await agent.invoke({ name: "Alice" });

  console.log(result); // Output: { text: "Hello, Alice!" }

  expect(result).toEqual({ text: "Hello, Alice!" });

  // #endregion example-function-agent-async-generator
});

test("Agent.shutdown", async () => {
  // #region example-agent-shutdown

  class MyAgent extends Agent {
    override process(input: Message): Message {
      return { text: `Hello, ${input}` };
    }

    override async shutdown() {
      console.log("Agent is shutting down...");
      // Clean up resources, close connections, etc.
    }
  }

  const agent = new MyAgent();

  const shutdown = spyOn(agent, "shutdown");

  await agent.shutdown();

  expect(shutdown).toHaveBeenCalled();

  // #endregion example-agent-shutdown
});

test("Agent.shutdown by `using` statement", async () => {
  // #region example-agent-shutdown-by-using

  class MyAgent extends Agent {
    override process(input: Message): Message {
      return { text: `Hello, ${input}` };
    }

    override async shutdown() {
      console.log("Agent is shutting down...");
      // Clean up resources, close connections, etc.
    }
  }

  // agent will be automatically disposed of at the end of this block
  await using agent = new MyAgent();

  const shutdown = spyOn(agent, "shutdown");
  expect(shutdown).not.toHaveBeenCalled();

  // #endregion example-agent-shutdown-by-using
});

test("Agent should return the original error from guide rails", async () => {
  const legalModel = new OpenAIChatModel();

  const financial = AIAgent.from({
    ...guideRailAgentOptions,
    model: legalModel,
    instructions: `You are a financial assistant. You must ensure that you do not provide cryptocurrency price predictions or forecasts.
<user-input>
{{input}}
</user-input>

<agent-output>
{{output}}
</agent-output>
`,
  });

  class MyAgent extends Agent {
    override process(_input: Message): Message {
      return {
        text: "Bitcoin will likely reach $100,000 by next month based on current market trends.",
      };
    }
  }

  const agent = new MyAgent({ guideRails: [financial] });

  spyOn(legalModel, "process").mockReturnValueOnce(
    Promise.resolve({
      json: {
        abort: true,
        reason:
          "I cannot provide cryptocurrency price predictions as they are speculative and potentially misleading.",
      },
    }),
  );

  const result = await agent.invoke({ message: "What will be the price of Bitcoin next year?" });

  expect(result).toEqual({
    $status: "GuideRailError",
    abort: true,
    reason:
      "I cannot provide cryptocurrency price predictions as they are speculative and potentially misleading.",
  });
});

test("Agent can be intercepted by guide rails", async () => {
  // #region example-agent-guide-rails

  const model = new OpenAIChatModel();

  const legalModel = new OpenAIChatModel();

  const aigne = new AIGNE({ model });

  const financial = AIAgent.from({
    ...guideRailAgentOptions,
    model: legalModel,
    instructions: `You are a financial assistant. You must ensure that you do not provide cryptocurrency price predictions or forecasts.
<user-input>
{{input}}
</user-input>

<agent-output>
{{output}}
</agent-output>
`,
  });

  const agent = AIAgent.from({
    guideRails: [financial],
  });

  spyOn(model, "process").mockReturnValueOnce(
    Promise.resolve({
      text: "Bitcoin will likely reach $100,000 by next month based on current market trends.",
    }),
  );

  spyOn(legalModel, "process").mockReturnValueOnce(
    Promise.resolve({
      json: {
        abort: true,
        reason:
          "I cannot provide cryptocurrency price predictions as they are speculative and potentially misleading.",
      },
    }),
  );

  const result = await aigne.invoke(agent, {
    message: "What will be the price of Bitcoin next month?",
  });

  console.log(result);
  // Output:
  // {
  //   "$status": "GuideRailError",
  //   "message": "I cannot provide cryptocurrency price predictions as they are speculative and potentially misleading."
  // }

  expect(result).toEqual({
    $status: "GuideRailError",
    message:
      "I cannot provide cryptocurrency price predictions as they are speculative and potentially misleading.",
  });

  // #endregion example-agent-guide-rails
});

test("Agent should respond result if no any guide rails error", async () => {
  const model = new OpenAIChatModel();

  const legalModel = new OpenAIChatModel();

  const aigne = new AIGNE({ model });

  const financial = AIAgent.from({
    ...guideRailAgentOptions,
    model: legalModel,
    instructions: `You are a financial assistant. You must ensure that you do not provide cryptocurrency price predictions or forecasts.
<user-input>
{{input}}
</user-input>

<agent-output>
{{output}}
</agent-output>
`,
  });

  const agent = AIAgent.from({
    guideRails: [financial],
    inputKey: "message",
  });

  spyOn(model, "process").mockReturnValueOnce(
    Promise.resolve({
      text: "You can use AIGNE Framework create a useful agent!",
    }),
  );

  spyOn(legalModel, "process").mockReturnValueOnce(
    Promise.resolve({
      json: {
        abort: false,
        reason: "That is a normal response",
      },
    }),
  );

  const result = await aigne.invoke(agent, { message: "How to create an agent?" });

  expect(result).toEqual({
    message: "You can use AIGNE Framework create a useful agent!",
  });
});

test("Agent inspect should return it's name", async () => {
  expect(inspect(AIAgent.from({}))).toBe("AIAgent");

  expect(inspect(AIAgent.from({ name: "test_agent" }))).toBe("test_agent");
});

test("Agent should check input and output schema", async () => {
  const plus = FunctionAgent.from({
    name: "test-agent-plus",
    inputSchema: z.object({
      a: z.number(),
      b: z.number(),
    }),
    outputSchema: z.object({
      sum: z.number(),
    }),
    process: async (input) => {
      return {
        sum: input.a + input.b,
      };
    },
  });

  expect(
    plus.invoke({ a: "foo" as unknown as number, b: "bar" as unknown as number }),
  ).rejects.toThrow(
    "Agent test-agent-plus input check arguments error: a: Expected number, received string, b: Expected number, received string",
  );

  spyOn(plus, "_process").mockReturnValueOnce({
    sum: "3" as unknown as number,
  });
  expect(plus.invoke({ a: 1, b: 2 })).rejects.toThrow(
    "Agent test-agent-plus output check arguments error: sum: Expected number, received string",
  );
});

test("Agent hooks simple example", async () => {
  // #region example-agent-hooks

  const model = new OpenAIChatModel();

  const aigne = new AIGNE({ model });

  const weather = FunctionAgent.from({
    name: "weather",
    description: "Get the weather of a city",
    inputSchema: z.object({
      city: z.string(),
    }),
    outputSchema: z.object({
      temperature: z.number(),
    }),
    process: async (_input) => {
      return {
        temperature: 25,
      };
    },
  });

  const agent = AIAgent.from({
    hooks: {
      onStart: (event) => {
        console.log("Agent started:", event.input);
      },
      onEnd: (event) => {
        console.log("Agent ended:", event.input, event.output);
      },
      onSkillStart: (event) => {
        console.log(`Skill ${event.skill.name} started:`, event.input);
      },
      onSkillEnd: (event) => {
        console.log(`Skill ${event.skill.name} ended:`, event.input, event.output);
      },
    },
    skills: [weather],
    inputKey: "message",
  });

  const hooks = agent.hooks.at(0);
  assert(hooks);

  const onStart = spyOn(hooks, "onStart");
  const onEnd = spyOn(hooks, "onEnd");
  const onSkillStart = spyOn(hooks, "onSkillStart");
  const onSkillEnd = spyOn(hooks, "onSkillEnd");

  spyOn(model, "process")
    .mockReturnValueOnce({ toolCalls: [createToolCallResponse("weather", { city: "Paris" })] })
    .mockReturnValueOnce({ text: "The weather in Paris is 25 degrees." });

  const result = await aigne.invoke(agent, { message: "What is the weather in Paris?" });

  console.log(result);
  // Output: { message: "The weather in Paris is 25 degrees." }

  expect(result).toEqual({ message: "The weather in Paris is 25 degrees." });
  expect(onStart).toHaveBeenLastCalledWith(
    expect.objectContaining({ input: { message: "What is the weather in Paris?" } }),
  );
  expect(onEnd).toHaveBeenLastCalledWith(
    expect.objectContaining({
      input: { message: "What is the weather in Paris?" },
      output: { message: "The weather in Paris is 25 degrees." },
    }),
  );
  expect(onSkillStart).toHaveBeenLastCalledWith(
    expect.objectContaining({
      input: expect.objectContaining({ city: "Paris" }),
      skill: weather,
    }),
  );
  expect(onSkillEnd).toHaveBeenLastCalledWith(
    expect.objectContaining({
      input: expect.objectContaining({ city: "Paris" }),
      output: { temperature: 25 },
      skill: weather,
    }),
  );

  // #endregion example-agent-hooks
});

test("Agent hook onHandoff should work correctly", async () => {
  const model = new OpenAIChatModel();

  const aigne = new AIGNE({ model });

  const triage = AIAgent.from({
    hooks: {
      onHandoff(event) {
        console.log("Agent handoff:", event.target.name);
      },
    },
    skills: [
      function transferToFeedback() {
        return feedback;
      },
    ],
    toolChoice: AIAgentToolChoice.router,
    inputKey: "message",
  });

  const hooks = triage.hooks.at(0);
  assert(hooks);

  const onHandoff = spyOn(hooks, "onHandoff");

  const feedback = AIAgent.from({
    inputKey: "message",
  });

  spyOn(model, "process")
    .mockReturnValueOnce({ toolCalls: [createToolCallResponse("transferToFeedback", {})] })
    .mockReturnValueOnce({ text: "Hello, I am feedback agent." });

  const result = await aigne.invoke(triage, { message: "I want to give feedback" });

  expect(onHandoff).toHaveBeenLastCalledWith(
    expect.objectContaining({
      source: triage,
      target: feedback,
      input: { message: "I want to give feedback" },
    }),
  );

  expect(result).toEqual({ message: "Hello, I am feedback agent." });
});

test("AgentInput/AgentOutput should infer correct types", async () => {
  const agent = AIAgent.from({
    inputSchema: z.object({
      name: z.string(),
      age: z.number(),
    }),
    outputSchema: z.object({
      greeting: z.string(),
      ageInMonths: z.number(),
    }),
  });

  expectType<AgentInput<typeof agent>>().is<{ name: string; age: number }>();
  expectType<AgentOutput<typeof agent>>().is<{ greeting: string; ageInMonths: number }>();
});

test("Agent must return a record type", async () => {
  // biome-ignore lint/security/noGlobalEval: just for testing
  const agent = FunctionAgent.from(() => eval("4 * 4"));

  expect(agent.invoke({}, { streaming: false })).rejects.toThrow(
    "expect to return a record type such as {result: ...}, but got (number): 16",
  );

  expect(readableStreamToArray(await agent.invoke({}, { streaming: true }))).rejects.toThrow(
    "expect to return a record type such as {result: ...}, but got (number): 16",
  );
});

test("Agent should merge default input before invoking", async () => {
  const agent = FunctionAgent.from({
    inputSchema: z.object({
      title: z.string(),
      description: z.string().optional(),
      foo: z.string().optional(),
    }),
    defaultInput: {
      description: "Default description",
      foo: { $get: "title" },
    },
    process: async (input) => input,
  });

  const result = await agent.invoke({ title: "Test Title" });
  expect(result).toEqual({
    title: "Test Title",
    description: "Default description",
    foo: "Test Title",
  });
});

test("Agent should prioritize direct input over default input", async () => {
  const agent = FunctionAgent.from({
    inputSchema: z.object({
      title: z.string(),
      description: z.string().optional(),
      foo: z.string().optional(),
    }),
    defaultInput: {
      description: "Default description",
      foo: { $get: "title" },
    },
    process: async (input) => input,
  });

  const result = await agent.invoke({
    title: "Test Title",
    description: "Custom description",
    foo: "Custom Foo",
  });
  expect(result).toEqual({
    title: "Test Title",
    description: "Custom description",
    foo: "Custom Foo",
  });
});

test("Agent should render task title (string template) correctly", async () => {
  const agent = AIAgent.from({
    taskTitle: "Test Task for message: {{message}}",
  });

  expect(await agent.renderTaskTitle({ message: "Hello" })).toBe("Test Task for message: Hello");
});

test("Agent should render task title (function) correctly", async () => {
  const agent = AIAgent.from({
    taskTitle: (input) =>
      input.lang === "zh" ? `测试任务: {{message}}` : `Test Task for message: {{message}}`,
  });

  expect(await agent.renderTaskTitle({ lang: "zh", message: "Hello" })).toBe("测试任务: Hello");
  expect(await agent.renderTaskTitle({ lang: "en", message: "Hello" })).toBe(
    "Test Task for message: Hello",
  );
});
