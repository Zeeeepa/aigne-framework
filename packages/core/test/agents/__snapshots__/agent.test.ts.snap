// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Agent.invoke with streaming response 1`] = `
[
  "Hello",
  ",",
  " ",
  "How",
  " ",
  "can",
  " ",
  "I",
  " ",
  "assist",
  " ",
  "you",
  " ",
  "today",
  "?",
]
`;
